### ISSUE TITLE
Support copy-like behaviour for MailMover; synching (select) tags 1:1 back to IMAP.

### ISSUE THREAD

**Issue by MacGyverNL** [teythoon/afew#102]
Since space is cheap, I like to be able to synchronize back most notmuch-tags as IMAP folders. The easiest way to do that, for me, is by having a flat folder structure, one folder per tag, and then simply copy the mail into each folder that corresponds to a tag. I tried to get this behaviour by using a MailMover configuration like `INBOX = 'tag:inbox':INBOX 'tag:test':test`, but this tries to move the mail on top of the existing one (crashing afew) and would probably still remove it afterwards even if the move failed gracefully.

What I would like is to be able to target the current folder, as above, which would then simply result in the mail not being removed after copying it to any other destinations. Is this a feasible feature?

The next step could be a general copy-feature which could work as follows. Specify a list of tags to apply this to. afew could then, for each tag, do a notmuch-query that finds messages tagged with that tag but not present in the folder with the corresponding name (could also do this with tag:folder-pairs). For each of these messages, find a source file to copy from (possibly need to iterate over all source files if it has more than one, to account for removals from IMAP), and copy it into the destination directory. Then, turn it around: for each tag, do a notmuch-query that finds messages present in the folder with the corresponding name but not tagged with that tag. Remove these messages. (This could also be done in FolderNameFilter.)

Now perform a notmuch new to detect the additions / removals (or perhaps do it directly in the db for each message, might be more reliable w.r.t. synchronization issues).

I've tried my hand at implementing a Python script to do only this a few times now, and each time I get bogged down in details of "what happens if IMAP synchronization happens before this / after this". Perhaps we can solve this in afew.

**Comment by agenbite** [comment]
I'd really love to have that functionality. Both ways folder synchronization is what prevents me from jumping to text-based email (since I can't entirely leave my old folder-based setup).

**Comment by GuillaumeSeren** [comment]
Hello !

@MacGyverNL ,

> INBOX = 'tag:inbox':INBOX 'tag:test':test, but this tries to move the mail on top of the existing one (crashing afew)

Ah yes this would require some kind of patch to work, you probably run into https://github.com/afewmail/afew/blob/master/afew/MailMover.py#L80

> What I would like is to be able to target the current folder, as above, which would then simply result in the mail not being removed after copying it to any other destinations. Is this a feasible feature?

Yes

> The next step could be a general copy-feature which could work as follows. Specify a list of tags to apply this to. afew could then, for each tag, do a notmuch-query that finds messages tagged with that tag but not present in the folder with the corresponding name (could also do this with tag:folder-pairs)

Like the Mover but whitout the deletion, the thing is if you do that,
you will end with multiple files corresponding to a mail, sync and all files based operation will be bigger.

> Perhaps we can solve this in afew.

As you describe it, I think it could a clone of the MailMover or a patch on it,
not really sure.

Feel free to open a Pull-Request on this subjet.

**Comment by GuillaumeSeren** [comment]
Hey @agenbite 

> I'd really love to have that functionality. Both ways folder synchronization is what prevents me from jumping to text-based email (since I can't entirely leave my old folder-based setup).

Both way folder sync is not handled by afew itself it is done by mbsync for example,
or I maybe you mean something else ?

**Comment by agenbite** [comment]
Yeah, you're right, sorry. It's not an issue for afew, but maybe an issue in which afew could help. 

I'd like to see seamless two-way synchronisation between my historic complicated IMAP folder tree and notmuch/afew, so I can use the latter in my usual computer without losing the information so far contained in the folder hierarchy, and still be able to access mail from the web anywhere (when my computer is not at hand).

I've been looking into that with no success so far. Any hint will be appreciated.

**Comment by GuillaumeSeren** [comment]
@agenbite Yes basically the way imap sync stuff force you into a specific maildir setup.

Maybe if you want to talk a bit more about this 'historic complicated IMAP folder tree' you can open you own issue ;)

**Comment by agenbite** [comment]
I found a pretty nice solution through [aerc](aerc-gmail.org). I can now combine folders and the search power of notmuch. Problem (more or less) solved. :)
